source/task_scheduler.c

```diff
@@ -15,173 +15,116 @@
 
 #include <aws/common/task_scheduler.h>
 
-#include <aws/common/priority_queue.h>
-#include <aws/common/thread.h>
-
 #include <assert.h>
 
 static const size_t DEFAULT_QUEUE_SIZE = 7;
 
-struct task_container {
-    uint64_t timestamp;
-    struct aws_task task;
-};
-
 static int s_compare_timestamps(const void *a, const void *b) {
-    uint64_t a_time = ((struct task_container *)a)->timestamp;
-    uint64_t b_time = ((struct task_container *)b)->timestamp;
+    uint64_t a_time = (*(struct aws_task **)a)->timestamp;
+    uint64_t b_time = (*(struct aws_task **)b)->timestamp;
     return a_time > b_time; /* min-heap */
 }
 
-static inline int s_get_next_task(
-    struct aws_task_scheduler *scheduler,
-    struct aws_task *task,
-    uint64_t run_before,
-    uint64_t *next_run_time);
+static void s_run_all(struct aws_task_scheduler *scheduler, uint64_t current_time, enum aws_task_status status);
+
+int aws_task_scheduler_init(struct aws_task_scheduler *scheduler, struct aws_allocator *alloc) {
+    assert(alloc);
```

Comment:
it would crash on the next line: `scheduler->alloc = alloc;`

my internal rule thus far has been not to check stuff like that if it's going to crash due to dereference in the 1st few lines of the function. ex: `struct thing_impl *impl = thing->impl;` I wouldn't `assert(thing)` since it's immediately accessed